lib/java/persistance/src/test/java/org/enso/persist/PersistanceTest.java

```diff
@@ -171,5 +173,84 @@ public record IntegerSupply(Supplier<Integer> supply) {}
   // @start region="self-annotation"
   @Persistable(id = 432436)
   public record ServiceSupply(Service supply) {}
+
   // @end region="self-annotation"
+
+  @Persistable(id = 432437)
+  public static class SelfLoop {
+    public Persistance.Reference<SelfLoop> self;
+
+    public Persistance.Reference<SelfLoop> self() {
+      return self;
+    }
+
+    public SelfLoop(Persistance.Reference<SelfLoop> self) {
+      this.self = self;
+    }
+  }
+
+  @Test
+  public void testReferenceLoopsInPersistance() throws Exception {
+    var obj = new SelfLoop(null);
+    // make the loop
+    obj.self = Persistance.Reference.of(obj);
+
+    var loaded = serde(SelfLoop.class, obj, -1);
+    assertSame(
+        "The recreated object again points to itself", loaded, loaded.self.get(SelfLoop.class));
+  }
+
+  @Persistable(id = 432439)
+  public record LongerLoop1(int x, Persistance.Reference<LongerLoop2> y) {
+    public LongerLoop1 {
+      new Exception("x: " + x + " y: " + y).printStackTrace();
+    }
+  }
+
+  @Persistable(id = 432440)
+  public record LongerLoop2(Persistance.Reference<LongerLoop3> y) {}
+
+  @Persistable(id = 432441)
+  public static class LongerLoop3 {
+    public final String a;
+    public Persistance.Reference<LongerLoop1> y;
+
+    public String a() {
+      return a;
+    }
+
+    public Persistance.Reference<LongerLoop1> y() {
+      return y;
+    }
+
+    public LongerLoop3(String a, Persistance.Reference<LongerLoop1> y) {
+      this.a = a;
+      this.y = y;
+    }
+  }
+
+  @Test
+  public void testLoopsBetweenDifferentTypes() throws Exception {
+    var obj3 = new LongerLoop3("a", null);
+    var obj2 = new LongerLoop2(Persistance.Reference.of(obj3));
+    var obj1 = new LongerLoop1(1, Persistance.Reference.of(obj2));
+    obj3.y = Persistance.Reference.of(obj1);
+
+    var loaded1 = serde(LongerLoop1.class, obj1, -1);
+    var r2 = loaded1.y().get(LongerLoop2.class);
+    var r3 = r2.y().get(LongerLoop3.class);
+    var r1 = r3.y().get(LongerLoop1.class);
+
+    assertFalse(
+        "Currently the loaded1 reference isn't shared with r3.y() reference", loaded1 == r1);
```

Comment:
The current behavior isn't ideal, but it is unlikely to happen in practice. It is unlikely the root of the deserialized graph would be part of the loop and for references inside of deserialized graph the cycle works well as the `for` loop below demonstrates.

Reply:
The condition is reverted since 03494b8 - the root object can now be part of the `Reference` cycle.